app/src/main/java/org/kiwix/kiwixmobile/webserver/ZimHostActivity.java

```diff
@@ -0,0 +1,471 @@
+
+    serviceConnection = new ServiceConnection() {
+
+      @Override
+      public void onServiceConnected(ComponentName className, IBinder service) {
+        HotspotService.HotspotBinder binder = (HotspotService.HotspotBinder) service;
+        hotspotService = binder.getService();
+        hotspotService.registerCallBack(ZimHostActivity.this);
+      }
+
+      @Override
+      public void onServiceDisconnected(ComponentName arg0) {
+      }
+    };
+
+    startServerButton.setOnClickListener(v -> {
+      //Get the path of ZIMs user has selected
+      if (!ServerUtils.isServerStarted) {
+        getSelectedBooksPath();
+        if (selectedBooksPath.size() > 0) {
+          startHotspotHelper();
+        } else {
+          Toast.makeText(ZimHostActivity.this, R.string.no_books_selected_toast_message,
+              Toast.LENGTH_SHORT).show();
+        }
+      } else {
+        startHotspotHelper();
+      }
+    });
+    locationServicesHelper = new LocationServicesHelper(ZimHostActivity.this);
+  }
+
+  private void startHotspotHelper() {
+    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+      toggleHotspot();
+    } else {
+      //if (isMobileDataEnabled(context)) {
+      //  mobileDataDialog();
+      //} else {
+      if (ServerUtils.isServerStarted) {
+        startService(createHotspotIntent(ACTION_STOP_SERVER));
+      } else {
+        startHotspotManuallyDialog();
+      }
+      //}
+    }
+  }
+
+  private void getSelectedBooksPath() {
+    for (BooksOnDiskListItem item : booksAdapter.getItems()) {
+      if (item.isSelected()) {
+        BooksOnDiskListItem.BookOnDisk bookOnDisk = (BooksOnDiskListItem.BookOnDisk) item;
+        File file = bookOnDisk.getFile();
+        selectedBooksPath.add(file.getAbsolutePath());
+        Log.v(TAG, "ZIM PATH : " + file.getAbsolutePath());
+      }
+    }
+  }
+
+  private void select(@NonNull BooksOnDiskListItem.BookOnDisk bookOnDisk) {
+    ArrayList<BooksOnDiskListItem> booksList = new ArrayList<>();
+    for (BooksOnDiskListItem item : booksAdapter.getItems()) {
+      if (item.equals(bookOnDisk)) {
+        item.setSelected(!item.isSelected());
+      }
+      booksList.add(item);
+    }
+    booksAdapter.setItems(booksList);
+  }
+
+  @Override protected void onStart() {
+    super.onStart();
+    bindService();
+  }
+
+  @Override protected void onStop() {
+    super.onStop();
+    unbindService();
+  }
+
+  private void bindService() {
+    bindService(new Intent(this, HotspotService.class), serviceConnection,
+        Context.BIND_AUTO_CREATE);
+  }
+
+  private void unbindService() {
+    if (hotspotService != null) {
+      unbindService(serviceConnection);
+    }
+  }
+
+  private void toggleHotspot() {
+    //Check if location permissions are granted
+    if (ContextCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION)
+        == PackageManager.PERMISSION_GRANTED) {
+
+      startService(createHotspotIntent(
```

Comment:
this action is really weird, you call the action `isEnabled` but it actually toggles the hotspot on/off, also it makes ZimHostActivity do work that the Service should be doing. How it is currently

```
case ACTION_IS_HOTSPOT_ENABLED:
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
          zimHostCallbacks.onHotspotStateReceived(hotspotManager.checkHotspotState());
        }
        break;

  @Override public void onHotspotStateReceived(@NonNull Boolean isHotspotEnabled) {
    if (isHotspotEnabled) //if hotspot is already enabled, turn it off.
    {
      startService(createHotspotIntent(ACTION_TURN_OFF_AFTER_O));
    } else //If hotspot is not already enabled, then turn it on.
    {
      locationServicesHelper.setupLocationServices();
    }
  }
```

How it should be
```
case ACTION_TOGGLE_HOTSPOT:
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
           if(hotspotManager.isHotspotEnabled()){
              stopHotspotAndDismissNotification();
           }
            else{
               zimHostCallbacks.turnOnHotspot();
           }
        }
        break;

  @Override public void turnOnHotspot() {
      locationServicesHelper.setupLocationServices()
  }

Reply:
actually instead of `turnOnHotSpot` which is quite misleading prefer `requestLocationAccess` and consequently rename `ACTION_TURN_ON_AFTER_O` to `ACTION_LOCATION_ACCESS_GRANTED`

Reply:
and we can improve `locationServicesHelper.setupLocationServices()`

change the signature to be   `public void setupLocationServices(LocationCallback callback)`

and you can write
```
 locationServicesHelper.setupLocationServices( () ->
 startService(createHotspotIntent(ACTION_LOCATION_ACCESS_GRANTED));
```

and you don't have to `implement LocationCallback` in the class definition of

Reply:
Fixed. b2870ddc02246f67d6fa09afca1400c226c53608